File: examples/python/s3_download.py
"""Descarga un objeto de S3 a un archivo local.

Uso: python3 s3_download.py <bucket> <clave> [ruta-destino]
"""
import sys

import boto3
from botocore.exceptions import ClientError

s3 = boto3.client(
    "s3",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)


def main():
    if len(sys.argv) < 3:
        raise SystemExit("Uso: python3 s3_download.py <bucket> <clave> [ruta-destino]")
    bucket, key = sys.argv[1], sys.argv[2]
    output = sys.argv[3] if len(sys.argv) > 3 else key.split("/")[-1]

    try:
        s3.download_file(bucket, key, output)
    except ClientError as error:
        if error.response["Error"]["Code"] == "404":
            raise SystemExit(f'No existe el objeto "{key}" en el bucket "{bucket}".')
        raise

    print(f"Descargado: s3://{bucket}/{key} -> {output}")


if __name__ == "__main__":
    main()

File: examples/go/s3_download.go
// Descarga un objeto de S3 a un archivo local.
// Uso: go run s3_download.go <bucket> <clave> [ruta-destino]
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func main() {
	if len(os.Args) < 3 {
		log.Fatal("Uso: go run s3_download.go <bucket> <clave> [ruta-destino]")
	}
	bucket := os.Args[1]
	key := os.Args[2]
	dest := filepath.Base(key)
	if len(os.Args) > 3 {
		dest = os.Args[3]
	}

	ctx := context.Background()
	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithRegion("us-east-1"),
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider("test", "test", "")),
	)
	if err != nil {
		log.Fatalf("Error cargando configuracion: %v", err)
	}
	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = true
		o.BaseEndpoint = aws.String("http://localhost:4566")
	})

	result, err := client.GetObject(ctx, &s3.GetObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
	if err != nil {
		log.Fatalf("Error descargando (¿existe %q en %q?): %v", key, bucket, err)
	}
	defer result.Body.Close()

	data, err := io.ReadAll(result.Body)
	if err != nil {
		log.Fatalf("Error leyendo el contenido: %v", err)
	}
	if err := os.WriteFile(dest, data, 0o644); err != nil {
		log.Fatalf("Error escribiendo %q: %v", dest, err)
	}
	fmt.Printf("Descargado: s3://%s/%s -> %s (%d bytes)\n", bucket, key, dest, len(data))
}
